### DISCUSSION TITLE
How to focus summary window on toggle?

### DISCUSSION THREAD

**Discussion by lawrence-laz** [nvim-neotest/neotest#197] [Q&A] [answered]
I use `lua require("neotest").summary.toggle()` to open and close summary window, but the window does not get focused automatically. Any way I could get it to focus on open?

**Comment by rcarriga** [comment, marked as answer]
You can do this in your mapping callback
```lua
      neotest.summary.toggle()
      local win = vim.fn.bufwinid("Neotest Summary")
      if win > -1 then
        vim.api.nvim_set_current_win(win)
      end
```

**Reply by lawrence-laz** [reply]
Great, this is what I was looking for!

It works just as expected, but LSP  is showing a warning on `bufwinid` call:
![image](https://user-images.githubusercontent.com/8823448/214542577-5b350d51-a630-43b7-a738-7ec6612813c0.png)

Any idea why this might happen?

**Reply by rcarriga** [reply]
Just incorrect type annotations from neodev.lua, perfectly safe to ignore

**Reply by guruor** [reply]
Actual summary toggle method expands the summary window till the nearest test, but with above snippet it always remains at the root of hierarchy.

Default behaviour:
<img width="398" alt="Screenshot 2023-04-23 at 6 49 33 AM" src="https://user-images.githubusercontent.com/14184045/233814166-4e949938-3058-4882-870a-615fccdb94b0.png">

Snippet behaviour:
<img width="328" alt="Screenshot 2023-04-23 at 6 49 10 AM" src="https://user-images.githubusercontent.com/14184045/233814164-833c159e-c854-458a-a1c3-4fad866b51a9.png">

**Reply by hjdivad** [reply]
@G0V1NDS 

```lua
          require("neotest").summary.toggle()
          vim.defer_fn(function()
            ---@diagnostic disable-next-line: param-type-mismatch
            local win = vim.fn.bufwinid("Neotest Summary")
            if win > -1 then
              vim.api.nvim_set_current_win(win)
            end
          end, 101)
```

Obviously just waiting 101ms is a bad workaround, but I couldn't quickly find a good hook or event to use.  The wait time is selected to outwait the [initial sleep](https://github.com/nvim-neotest/neotest/blob/dcdb40ea48f9c7b67a5576f6bb2e5f63ec15f2c0/lua/neotest/consumers/summary/summary.lua#L129-L130) as well as because that's what worked for me.

You still don't get best results because the cursor will be on the wrong line in the summary panel, but you should see it expanded at least.

**Reply by Terseus** [reply]
Hey there @guruor @hjdivad 

Here's a solution that doesn't use a hard coded sleep time and select correctly the focused test:

```lua
            vim.api.nvim_create_autocmd("User", {
                pattern = "NeotestSummaryOpen",
                callback = function()
                    vim.defer_fn(
                        function() vim.api.nvim_set_current_win(vim.fn.bufwinid("Neotest Summary")) end,
                        1
                    )
                end,
            })
```

The "NeotestSummaryOpen" user event pattern can be found [here](https://github.com/nvim-neotest/neotest/blob/f30bab1faef13d47f3905e065215c96a42d075ad/lua/neotest/consumers/summary/summary.lua#L42), now if we don't apply a `defer_fn` with 1ms of delay it won't work, will select the root element.

No idea why the 1ms delay works, but it works :slightly_smiling_face:
